### PR TITLE
Provide kubernetes configuration

### PR DESCRIPTION
**Pull request by viveksyngh**
This changes provider kubernetes configuration for of-builder and
registry as .yml file in yaml folder.

Also replaces hardcoded gateway URL with environment variable
gateway_url.

Fixes: #1

Signed-off-by: Vivek Singh <vivekkmr45@yahoo.in>

### CODE REVIEW

**Comment by alexellis** [comment]
Hi Vivek, thanks for raising this PR. Have you been able to test the changes too?

**Comment by CoderAway28th** [comment]
The namespaces are unmatched. @viveksyngh . Of_builder and registry namespace is "openfaas"
while gh-push and git-tar .. namespace is "openfaas-fn". That may cause nslookup error if you don't specify namespace for of_builder and registry urls.

**Comment by alexellis** [comment]
Can we change point 1 so that it uses a unix timestamp instead?

**Comment by viveksyngh** [comment]
@alexellis This is ready to merge along with #25 . Both has been tested on Docker swarm and k8s.

**Comment by alexellis** [comment]
OK this is on my list to review.. if anyone else wants to try it out that'd help too.

**Comment by alexellis** [comment]
cc @affix @johnmccabe

**Comment by affix** [comment]
@alexellis @johnmccabe @viveksyngh do we really need to expose these services to the network?

If they are all called from inside the k8s namespace and are not supposed to have external comms then we shouldn't publish the services.

A ReplicaSet instead of a Service may be more suited to the builders as it makes it easier to scale in and out without exposing the services publically.

File: yaml/of_builder.yml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  namespace: openfaas
+  labels:
+    app: of-builder
+spec:
+  type: NodePort
+  ports:
+    - port: 8080
+      protocol: TCP
+      targetPort: 8080
+      nodePort: 31121
```
Comment (alexellis) [inline yaml/of_builder.yml]: The nodePort can be removed, it'd only be needed for testing the of-builder manually. Maybe we can use a [ClusterIP](https://kubernetes.io/docs/concepts/services-networking/service/) instead?

File: yaml/registry.yml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  namespace: openfaas
+  labels:
+    app: registry
+spec:
+  type: NodePort
+  ports:
+    - port: 5000
+      protocol: TCP
+      targetPort: 5000
+      nodePort: 31115
```
Comment (alexellis) [inline yaml/registry.yml]: @affix I think we could push from the of-builder/of-buildkit service but not pull from it - the Kubelet wouldn't deploy from a ClusterIP

File: yaml/of-buildkit.yml [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+          privileged: true
+---
+apiVersion: v1
+kind: Service
+metadata:
+  name: of-buildkit
+  namespace: openfaas
+  labels:
+    app: of-buildkit
+spec:
```
Comment (alexellis) [inline yaml/of-buildkit.yml]: This NodePort can be removed in favour of a ClusterIP. I've also updated the README file for Swarm.

**Comment by alexellis** [comment]
Let's merge this and then have someone with Kubernetes experience give it a once-over. Thanks for getting us this far @viveksyngh 👍

**Comment by affix** [comment]
i'll deploy it tonight @alexellis (will stick it on the same cluster as my normal faas deployment in a different namespace)

**Comment by viveksyngh** [comment]
@alexellis I had updated this to used `ClusterIP` instead of `NodePort`.  @affix Let me know if you face any issue in running this. There are few things that need to be configured in a different way for kubernetes, which I have listed in previous comments.

**Comment by alexellis** [comment]
Lots of config changes are needed in the yaml files where we refer to `gateway` it must be `gateway.openfaas` etc

**Pull request merged**